libraries/lib-concurrency/ParallelProgress.h

```diff
@@ -0,0 +1,55 @@
+/**********************************************************************
+ 
+ Audacity: A Digital Audio Editor
+ 
+ @file ParallelProgress.h
+ 
+ Paul Licameli
+ 
+ **********************************************************************/
+#ifndef __AUDACITY_PARALLEL_PROGRESS__
+#define __AUDACITY_PARALLEL_PROGRESS__
+
+#include <atomic>
+
+namespace Parallel {
+
+namespace detail { struct AbortException{}; }
+
+//! A callback supplied to run methods of a task so it can "yield" while it
+//! reports progress, also detecting total progress
+/*!
+ Throws an exception when a thread team is shutting down.  Do not handle this
+ exception.
+
+ @tparam Increment an Integral type
+ */
+template<typename Increment> class Progress {
+public:
+   using increment_type = Increment;
+
+   //! Report progress; poll total progress; possibly non-returning
+   /*!
+    This may throw an exception internally to stop a thread when it has been
+    detected that other threads in the team have stopped.
+
+    TODO C++20: make Progress an awaitable that can suspend instead of throwing,
+    and let tasks be coroutines
+    */
+   increment_type operator()(increment_type inc = {}) {
```

Comment:
What is the justification for making progress callable? I would not object if the users below were allowed to receive a callable instead of progress. However, Progress is used only to stop the workers if the exception occurs.